File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.java
import java.util.*;

class Solution {
    public int[] solution(int[] numbers) {
        
        Deque<Integer> que = new ArrayDeque<>();
        int answer[] = new int[numbers.length];
        
        for(int i=numbers.length-1; i>=0; i--){
            while(!que.isEmpty()){
                if(que.peekLast() <= numbers[i]) que.pollLast();
                else break;
            }
            if(que.isEmpty()){
                answer[i] = -1;
            }else answer[i] = que.peekLast();
            que.add(numbers[i]);
        }

        return answer;
    }
}

File: 프로그래머스/2/154539. 뒤에 있는 큰 수 찾기/뒤에 있는 큰 수 찾기.js
function solution(numbers) {
    
    const que = [];
    const answer = Array(numbers.length).fill(0);
    
    for(let i=numbers.length-1; i>=0; i--){
        while(que.length){
            if(que[que.length-1] <= numbers[i]) que.pop();
            else break;
        }
        if(que.length){
            answer[i] = que[que.length-1]
        }else answer[i] = -1
        que.push(numbers[i])
    }
  
    return answer;
}